File: monitors/philosophers.py
#! /usr/bin/env python

import threading
import time

PHILOSOPHERS = 5
EAT_COUNT = 100

class Philosopher(threading.Thread):
    mutex = threading.Lock()
    forks = [] #forks available for each philosopher
    canEat = []
    count = 0

    def __init__(self):
        super(Philosopher, self).__init__()
        self.id = Philosopher.count
        self.right = (self.id - 1) % PHILOSOPHERS
        self.left = (self.id + 1) % PHILOSOPHERS
        Philosopher.count += 1
        Philosopher.forks.append(2)
        Philosopher.canEat.append(threading.Condition(Philosopher.mutex))

    def pick(self):
        with Philosopher.mutex:
            while Philosopher.forks[self.id] != 2:
                Philosopher.canEat[self.id].wait()

            Philosopher.forks[self.left] -= 1
            Philosopher.forks[self.right] -= 1

    def release(self):
        with Philosopher.mutex:
            Philosopher.forks[self.left] += 1
            Philosopher.forks[self.right] += 1

            if Philosopher.forks[self.left] == 2:
                Philosopher.canEat[self.left].notify()
            if Philosopher.forks[self.right] == 2:
                Philosopher.canEat[self.right].notify()

    def think(self):
        time.sleep(0.05)

    def eat(self):
        print("{} start eat".format(self.id))
        time.sleep(0.1)
        print("{} end eat".format(self.id))

    def run(self):
        for i in range(EAT_COUNT):
            self.think()
            self.pick()
            self.eat()
            self.release()

def main():

    philosophers = []

    for i in range(PHILOSOPHERS):
        philosophers.append(Philosopher())

    # Start all threads
    for p in philosophers:
        p.start()

    # Wait for all threads to complete
    for p in philosophers:
        p.join()


if __name__ == "__main__":
    main()

File: channels/philosophers.go
package main

import (
    "fmt"
    "runtime"
    "time"
)

const (
    GOROUTINES   = 10
    PHILOSOPHERS = 5
    EAT_COUNT    = 100
)

type Empty struct{}

var forks [PHILOSOPHERS]chan Empty

func think(id int) {
    time.Sleep(50 * time.Millisecond)
}

func pick(id int) {
    right := (id + 1) % PHILOSOPHERS
    if id < right {
        <-forks[id]
        <-forks[right]
    } else {
        <-forks[right]
        <-forks[id]
    }
}

func eat(id int) {
    fmt.Printf("%d start eat\n", id)
    time.Sleep(100 * time.Millisecond)
    fmt.Printf("%d end eat\n", id)
}

func release(id int) {
    right := (id + 1) % PHILOSOPHERS
    forks[id] <- Empty{}
    forks[right] <- Empty{}
}

func philosopher(id int, done chan Empty) {
    for i := 0; i < EAT_COUNT; i++ {
        think(id)
        pick(id)
        eat(id)
        release(id)
    }
    done <- Empty{}
}

func main() {
    runtime.GOMAXPROCS(GOROUTINES)
    done := make(chan Empty, 1)

    for i := range forks {
        forks[i] = make(chan Empty, 1)
        forks[i] <- Empty{}
    }

    for i := 0; i < PHILOSOPHERS; i++ {
        go philosopher(i, done)
    }

    for i := 0; i < PHILOSOPHERS; i++ {
        <-done
    }

    fmt.Printf("End\n")
}
